lib/opam_solve.ml

```diff
@@ -121,99 +158,261 @@ let fixed_packages ~local_packages ~pin_depends =
+      discared by [candidates]. *)
+  let create ?test ~constraints (env, paths) =
+    let env varname = String.Map.find_opt varname env in
+    match paths with
+    | [] ->
+        invalid_arg
+          "Multi_dir_context should be instanciated with at least one repo"
+    | _ ->
+        (* TODO: properly set the env *)
+        List.map ~f:(create ?test ~constraints ~env) paths
+
+  let merge_candidates acc candidates =
+    List.fold_left ~init:acc candidates ~f:(fun acc (version, opam_file_res) ->
+        OpamPackage.Version.Map.update version
+          (fun existing_opam -> existing_opam)
+          opam_file_res acc)
+
+  let candidates t name =
+    let map =
+      List.fold_left t ~init:OpamPackage.Version.Map.empty
+        ~f:(fun acc dir_context ->
+          let candidates = candidates dir_context name in
+          merge_candidates acc candidates)
+    in
+    OpamPackage.Version.Map.bindings map
+
+  let user_restrictions t pkg = user_restrictions (List.hd t) pkg
+
+  let filter_deps t pkg f = filter_deps (List.hd t) pkg f
+end
+
+module Local_opam_context : BASE_CONTEXT with type input = switch = struct
+  include Opam_0install.Switch_context
+
+  type input = OpamStateTypes.unlocked OpamStateTypes.switch_state
+
+  let create ?test ~constraints switch_state =
+    create ?test ~constraints switch_state
+end
+
+(* The code below aims to provide a unified interface over the two solver
+   modules *)
+
+type ('a, 'b) t =
+  (module OPAM_MONOREPO_SOLVER with type input = 'a and type diagnostics = 'b)
+
+module Local_opam_config_context = Opam_monorepo_context (Local_opam_context)
+module Explicit_repos_context = Opam_monorepo_context (Multi_dir_context)
+module Local_opam_config_solver = Make_solver (Local_opam_config_context)
+module Explicit_repos_solver = Make_solver (Explicit_repos_context)
+
+type switch_diagnostics = Local_opam_config_solver.diagnostics
+
+type explicit_repos_diagnostics = Explicit_repos_solver.diagnostics
+
+let local_opam_config_solver : (switch, switch_diagnostics) t =
+  (module Local_opam_config_solver)
+
+let explicit_repos_solver :
+    (opam_env * explicit_repos, explicit_repos_diagnostics) t =
+  (module Explicit_repos_solver)
+
+let calculate :
+    type context diagnostics.
+    build_only:bool ->
+    allow_jbuilder:bool ->
+    local_opam_files:
+      (OpamTypes.version * OpamFile.OPAM.t) OpamPackage.Name.Map.t ->
+    target_packages:OpamPackage.Name.Set.t ->
+    pin_depends:(OpamTypes.version * OpamFile.OPAM.t) OpamPackage.Name.Map.t ->
+    ?ocaml_version:string ->
+    (context, diagnostics) t ->
+    context ->
+    ( Opam.Package_summary.t list,
+      [> `Diagnostics of diagnostics | `Msg of string ] )
+    result =
+ fun ~build_only ~allow_jbuilder ~local_opam_files ~target_packages ~pin_depends
+     ?ocaml_version t input ->
+  let (module Solver : OPAM_MONOREPO_SOLVER
+        with type diagnostics = diagnostics
+         and type input = context) =
```

Comment:
Could you be a bit more specific, I genuinely did not understand what you were suggesting or what was wrong with this piece of code.

I agree it is fairly convoluted but given we have to different modules that's the best I could come up with and I think I managed to keep it contained in a few functions.

Reply:
My hang-up (if you want to call it this even) is that due to the fact that the module is passed in as argument, all the functions have to repeat this

```ocaml
(module Solver : OPAM_MONOREPO_SOLVER
        with type diagnostics = diagnostics
         and type input = context)
```

to reconstruct a module-language module out of it. I'm wondering whether there would be a way to restructure the code where this wouldn't be necessary? Maybe these functions can be replaced by directly calling them on the functor (`Local_opam_config_solver.calculate` instead of `calculate (module Local_opam_config_solver)`)?

Reply:
The goal is to abstract this module madness away from the external API of the module. I don't want to have as many code path as there implementations outside of `Opam_solve`. If you have a suggestion that solves that without being too messy I'll gladly replace what we have now but given each of the module define its own types, I don't see how this could be solved without first class modules or GADTs. Maybe the latter would look nicer?

Reply:
I've been thinking about whether there's a reasonable way to make this API nice for the caller but without the boilerplate of first-class modules but most things I have come up with actually move the boilerplate from the callee to the caller which I think is worse so I think this is fine.